### ISSUE TITLE
Run 1.2i preflight checklist

### ISSUE THREAD

**Issue by cwwalter** [LSSTDESC/imSim#100]
I'm copying over a list of items from a recent email thread to be finalized before the run 1.1i run starts so we can checkoff on them.  Please mark off the checks with the PRs are merged and comment further in this issue.

- [X] Sky model variation code from Emily and Bela (modified by Jim)
- [x] Craig’s tree ring code.
- [x] Francois’ morphology code
- [x] Josh’s atmospheric code
- [x] Daniel’s optical aberration Zernike representation of sensitivity matrix 
- [x] Jim's PR for multiprocessor code to leverage KNL in the pipeline

Two outstanding items from before got some study this week but might wait for r1.2:

- [x] Charge bleeding
- [x] Sky background tree rings through single surface op to increase speed.

Waiting for DC3:

- [x] Chromatic Seeing

**Comment by cwwalter** [comment]
@jmeyers314 following up on the SSim slack discussion today:  I had forgotten to followup on chromatic response.  We should decide what to do for run1.1.

**Comment by jmeyers314** [comment]
Do you mean chromatic seeing?  (i.e., lambda^-0.2 for pure Kolmogorov turbulence?).  I don't think there's an easy way for us to do that in GalSim right now.  We'd have to figure out a way to assign wavelengths and rescale the PSF kicks before convolving by the galaxy profile, which isn't really supported.  That will probably have to wait for DC3.

**Comment by cwwalter** [comment]
> Do you mean chromatic seeing? (i.e., lambda^-0.2 for pure Kolmogorov turbulence?).

Yes.  That's what I mean.

Hmm... OK.  Let me try to get a better understanding of the situation:

I thought if we just wanted the pure case like above (to start say) we could just scale the kicks like was done in the PhoSim scheme.  Then, something more complicated would need to utilize a more complicated function like a digitized version of what Ji Won worked out, or perhaps from that paper. Eventually I assume we would parameterize something carefully from the full DFT calculation.

Are you saying we don't know how to do that yet, or that what I wrote above is actually not the solution and we need to do something more complicated which we don't understand yet?  If that is the case is the 1st solution possible in the shorter term?

**Comment by jmeyers314** [comment]
>I thought if we just wanted the pure case like above (to start say) we could just scale the kicks like was done in the PhoSim scheme.

Yeah, I think this is the right approach too.  I'm just not sure there's a good way to do that right now using the GalSim API.  If all we had were stars, we could do it by adding a `rescale_the_atm_kicks_using_wavelength` `surface_op`.  But this breaks down for galaxies because we only want to rescale the atmospheric PSF kicks, and not the kicks that correspond to convolving by a galaxy surface brightness profile.

I'm sure we could come up with a dirty hack if we decide this is really important though.  I'm curious what @rmjarvis has to say...

**Comment by rmjarvis** [comment]
> I'm sure we could come up with a dirty hack if we decide this is really important though. I'm curious what @rmjarvis has to say...

I agree with Josh.  We don't have a good way to do that yet.  This needs to wait for the real chromatic photon shooting.  Doing it as a surface op isn't right, since this is after the convolution.  You really want a different PSF convolution for each photon.  There is no correct rescaling to do after you've already convolved by a monochromatic PSF.  

I think this just needs to be "not a feature" of the imsim DC2 runs.

**Comment by cwwalter** [comment]
> I agree with Josh. We don't have a good way to do that yet. This needs to wait for the real chromatic photon shooting.

OK, I understand.  So we will tie this to Chromatic Photon Shooting and this is a DC3 not DC2 era feature since this will require some non-trivial retooling and development.  I'll enter a separate issue and make a DC3 label.

**Comment by cwwalter** [comment]
> But this breaks down for galaxies because we only want to rescale the atmospheric PSF kicks, and not the kicks that correspond to convolving by a galaxy surface brightness profile.

Just trying to remind myself.. Is this because you included this  galaxy convolution in the same step so as to kill the rapidly changing function that gave the unphysical delta function before?  Or am I remembering something else?

**Comment by rmjarvis** [comment]
It's because the convolution happens before we pick wavelengths.  Reversing the order of these operations requires the chromatic photon shooting.  (Or you can use the chromatic FFT stuff, where all this is already implemented, but that's way too slow for DC2.)

**Comment by jmeyers314** [comment]
>Just trying to remind myself.. Is this because you included this galaxy convolution in the same step so as to kill the rapidly changing function that gave the unphysical delta function before? Or am I remembering something else?

It's because of what Mike said: the correct order of operations just isn't currently supported.

What you're thinking about is that I'm now including convolution by an Airy profile in the second kick calculation to help kill the wiggles present in a "pure" atmosphere-only second kick.

**Comment by cwwalter** [comment]
Great. Thanks both; I understand.

**Comment by cwwalter** [comment]
Noting that DCR has it's own issue in #31.  @rmjarvis is anything happening on the DCR surface operation front, or should this be moved to R1.2?

**Comment by rmjarvis** [comment]
> @rmjarvis is anything happening on the DCR surface operation front, or should this be moved to R1.2?

I'm working on it.  Also looking into whether I can significantly speed up the wavelength sampler.  I don't really understand the timings Jim is getting for that.  But nothing to report from my end yet.

**Comment by cwwalter** [comment]
OK; got it. Thanks Mike.

**Comment by rmjarvis** [comment]
Cross reference [GalSim PR for DCR surface op](https://github.com/GalSim-developers/GalSim/pull/961)

**Comment by cwwalter** [comment]
Note: GalSim DCR surface Op https://github.com/GalSim-developers/GalSim/pull/961 is now merged in GalSim. I believe it still will need to be activated on the imSim side.

**Comment by cwwalter** [comment]
Also there is much progress on sky background treerings and flats (including high intensity versions appropriate for BF correction).  We can now do this very quickly.  The code was implemented by Mike and the code is being reviewed and tested here: https://github.com/GalSim-developers/GalSim/pull/963.

**Comment by cwwalter** [comment]
@jmeyers314 @djperrefort @rmjarvis @jchiang87 : These three remaining items:

 - Josh’s atmospheric code (#101)
 - DCR GalSim surface operation (ready in GalSim needs to be turned on)
 -  Daniel’s optical aberration Zernike representation of sensitivity matrix (#102)

Are all ready to go but we need to merge them in. I understand we are waiting to do this because the Zernike code in GalSim has been merged to master but it is not in a release.  What is the release plan? I think these are the last things we need for DC2. Alternatively could we use GalSim master (or a branch) for testing now?

I think to turn DCR on we need to add it to the surface operations list in the photon shooting drawImage call in the interface code. @danielsf and @rmjarvis do you agree?

**Comment by rmjarvis** [comment]
I think from the GalSim side, we are all set.  We'll release version 1.6 with the current feature set probably this week or so. 

> I think to turn DCR on we need to add it to the surface operations list in the photon shooting drawImage call in the interface code.

Yes.  It should be trivial to add that as another surface op.  It needs to come after the wavelength one in the list.

**Comment by djperrefort** [comment]
Once the new GalSim version is released the Optical Zernike code will be set to go. Referring to a previous email conversation, it should be ready to interface with @jmeyers314's code as is.

**Comment by jchiang87** [comment]
Yes, we should move forward, but I'd like to go through the raytracedAtmosphere PR first.  I'll start on that tonight.

**Comment by cwwalter** [comment]
@danielsf are @EiffL's morphology knots already being produced for the instance catalogs? Or does something else have to happen to make sure we are/including testing that?

**Comment by EiffL** [comment]
The instance catalog needs to be produced using a protoDC2 knots addon catalog as input. The last one I produced was weeks ago, so I need to make sure the addon exists for the latest version of the protoDC2 catalog.

**Comment by EiffL** [comment]
Ok, updated, to create instance catalogs with knots, simply use `proto-dc2_v3.0_addon_knots` as the input GCR catalog.

**Comment by cwwalter** [comment]
Thanks.  @danielsf and @jchiang87 are you all set to use this or is there anything else on your side?

**Comment by jchiang87** [comment]
There's a [PR](https://github.com/LSSTDESC/sims_GCRCatSimInterface/pull/47) that I just made in `sims_GCRCatSimInterface` that we should merge first, but other than that, I think this is sufficient for simulating the knots in a Run1.1i using protoDC2 v3.0.

**Comment by cwwalter** [comment]
Was DCR turned on the surface effect list?

**Comment by jchiang87** [comment]
It wasn't....we should probably make an issue to add it at the `sims_GalSimInterface` repo.

**Comment by cwwalter** [comment]
> It wasn't....we should probably make an issue to add it at the sims_GalSimInterface repo.

OK I will do it. I thought there was an issue where @rmjarvis explicitly laid out what to do but I don't see it now..

**Comment by rmjarvis** [comment]
The only thing you need to know that might not be in the data you already have access to is the hour angle of the observation.  Then the usage looks like:
```
    dcr = galsim.PhotonDCR(base_wavelength = bandpass.effective_wavelength,
                           HA = observation_hour_angle,
                           latitude = site_latitude,
                           obj_coord = galsim.CelestialCoord(ra, dec))
    surface_ops = [wave_sampler, dcr] #, ... any others as well.
```
dcr just needs to be after the wavelength sampler in the list of surface ops.

**Comment by jchiang87** [comment]
One more item regarding position angles:
- [ ] Galaxy position angles in imSim are defined with PA increasing from zero eastward of NCP.  I *think* phosim defines PA as increasing from zero westward of NCP (@danielsf Can you confirm?).  If so, this also needs to be changed in order for the two sims to agree.  We should probably check that the extragalactic catalogs have the same convention as phosim for the galaxy PAs.

**Comment by rmjarvis** [comment]
What is the distinction between PA and shear convention?  Are you doing some calculation to go from the e1,e2 values in the input catalog to turn that in to |e|, PA?  And then applying this angle with the e, beta notation in GalSim?

It seems to me that we should just be taking the e1,e2 values from the input catalog and applying them directly, but maybe I'm missing something.  (And as already discussed, there is an E/W difference between PhoSim and ImSim wrt the use of e1,e2, which manifests as a sign difference in e2.)

**Comment by jchiang87** [comment]
For shears, we are just using g1, g2.  For (almost all) galaxy shapes, independent of any WL shears, we use the `sersic2D` model,and  the instance catalogs provide the parameters that phosim expects: major, minor axis half-light radii, position angle, and sersic index.  Both the WL shears and galaxy position angles have a sign and/or angle conventions that are currently different for phosim and imsim.

**Comment by cwwalter** [comment]
In appendix 3 of 

https://arxiv.org/abs/1705.06766

the convention for the HSC (and LSST pipeline) is defined as:

"and position angle (θ), which we define here as the angle from the x-axis to the semi-major axis, measured counterclockwise."

Can we make this correspondence to our variables and choose the convention that matches it?  Or is the this still unconstraining?

**Comment by rmjarvis** [comment]
> and position angle (θ), which we define here as the angle from the x-axis to the semi-major axis, measured counterclockwise."

The question I guess is, what is the x-axis?  The CCD x-axis?  If so, that's a terrible choice, since that will be different for every observation of a given galaxy, so that's probably not it.  Possibly, the x-axis of the coadd image makes sense.  That's probably what they used for that paper, but then we'd still need to know if the +x direction there is east or west.  (I.e. is there a reflection relative to how the scene is viewed from earth, since presumably +y would be north -- that's relatively standard.)

**Comment by cwwalter** [comment]
I changed the name of the issue to be 1.2i to reflect our new naming scheme (based on the instance catalog no the number of attempts).

**Comment by cwwalter** [comment]
I'm moving the DCR sub-task to the 2.0i preflight checklist.  All items on this list are now closed and production is starting in the UK.

So, I am closing this issue.
